<a href="https://colab.research.google.com/github/JayeshHarjule/LetsUpgradeAI-ML/blob/master/Day31.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Answer 1)**

**Problem Statement:** Use IRIS dataset to perform KNN and compare with logistic regression.

In [ ]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
import sklearn as sk
from sklearn import datasets
from sklearn.model_selection import train_test_split
from collections import Counter
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler 
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix  
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report
%matplotlib inline


iris = datasets.load_iris()

#separate x and y labels
training = iris['data']
training_labels = iris['target']

#create x and y dataframes
training_df = pd.DataFrame(training,columns=['sepal length','sepal width','petal length','petal width'])
labels_df = pd.DataFrame(training_labels,columns=['Iris Species'])

#join them together horizontally
training_df = pd.concat([training_df,labels_df],axis=1)

#separate x and y dataframes
labels_df = training_df['Iris Species']
training_df.drop('Iris Species',axis=1,inplace=True)

#create training and testing sets
x_train,x_test,y_train,y_test = train_test_split(training_df,labels_df,test_size=.33,random_state=0)

#Fitting K-NN classifier to the training set  
classifier= KNeighborsClassifier(n_neighbors=5, metric='minkowski', p=2 )  
classifier.fit(x_train, y_train)


#Predicting the test set result  
y_pred= classifier.predict(x_test)


#Creating the Confusion matrix  
cm1= confusion_matrix(y_test, y_pred)
print("Confusion Matrix when KNN Classifier is impelemented \n", cm1)
print("\n Accuracy of test set with KNN classifier =",accuracy_score(y_test, y_pred)*100)
print("\n The classification report when KNN classifier is applied is as follows \n")
print(classification_report(y_test, y_pred))


#feature Scaling  
st_x= StandardScaler()    
x_train= st_x.fit_transform(x_train)    
x_test= st_x.transform(x_test) 

classifier= LogisticRegression()

# fit the model with data
classifier.fit(x_train,y_train)

#Predicting the test set result
y_pred=classifier.predict(x_test)

#Creating the Confusion matrix  
cm2= confusion_matrix(y_test, y_pred)
print("\n Confusion Matrix when Logistic Regression is impelemented \n", cm2)
print("\n Accuracy of test set with Logistic Regression =",accuracy_score(y_test, y_pred)*100)
print("\n The classification report when Logistic Regression is applied is as follows \n")
print(classification_report(y_test, y_pred))

Confusion Matrix when KNN Classifier is impelemented 
 [[16  0  0]
 [ 0 18  1]
 [ 0  0 15]]

 Accuracy of test set with KNN classifier = 98.0

 The classification report when KNN classifier is applied is as follows 

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        16
           1       1.00      0.95      0.97        19
           2       0.94      1.00      0.97        15

    accuracy                           0.98        50
   macro avg       0.98      0.98      0.98        50
weighted avg       0.98      0.98      0.98        50


 Confusion Matrix when Logistic Regression is impelemented 
 [[16  0  0]
 [ 0 18  1]
 [ 0  2 13]]

 Accuracy of test set with Logistic Regression = 94.0

 The classification report when Logistic Regression is applied is as follows 

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        16
           1       0.90      0.95      0.92        19
     

Accuracy of KNN classifier is 98% whereas it is 94% when Logistic Regression is applied on IRIS dataset.